src/exodus/sigmawallet.cpp

```
@@ -0,0 +1,597 @@
+
+    if (!SetupWallet(masterId)) {
+
+        LogPrintf("%s: failed to save deterministic seed for hashseed %s\n", __func__, masterId.GetHex());
+        throw std::runtime_error("fail to setup wallet");
+    }
+}
+
+bool SigmaWallet::SetupWallet(const uint160& masterId)
+{
+    CWalletDB walletdb(walletFile);
+
+    if (pwalletMain->IsLocked()) {
+        return false;
+    }
+
+    if (masterId.IsNull()) {
+        return error("%s: failed to set master seed.", __func__);
+    }
+
+    this->masterId = masterId;
+
+    // Load mint pool from DB
+    LoadMintPool();
+
+    // Clean up any mint entry that isn't corresponded to current masterId
+    CleanUp();
+
+    // Refill mint pool
+    GenerateMintPool();
+
+    return true;
+}
+
+// Generator
+uint32_t SigmaWallet::CreateNextSeed(CKeyID &seedId, uint512& seed)
+{
+    LOCK(pwalletMain->cs_wallet);
+    seedId = pwalletMain->GenerateNewKey(BIP44_EXODUS_MINT_INDEX).GetID();
+    return GenerateSeed(seedId, seed);
+}
+
+uint32_t SigmaWallet::GenerateSeed(CKeyID const &seedId, uint512& seed)
+{
+    CKey key;
+    if (!pwalletMain->CCryptoKeyStore::GetKey(seedId, key)) {
+        throw std::runtime_error(
+            "Unable to retrieve generated key for mint seed. Is the wallet locked?");
+    }
+
+    // HMAC-SHA512(SHA256(count), key)
+    std::array<unsigned char, CSHA256::OUTPUT_SIZE> countHash;
+
+    std::vector<unsigned char> result;
+    result.resize(CSHA512::OUTPUT_SIZE);
+
+    auto count = std::to_string(GetSeedIndex(seedId));
+    CSHA256().
+        Write(
+            reinterpret_cast<const unsigned char*>(count.data()),
+            count.size()
+        ).
+        Finalize(countHash.data());
+
+    CHMAC_SHA512(countHash.data(), sizeof(countHash)).
+        Write(key.begin(), key.size()).
+        Finalize(result.data());
+
+    seed = uint512(result);
+
+    return GetSeedIndex(seedId);
+}
+
+bool GetChildFromPath(string const &path, std::pair<uint32_t, bool> &child)
```

We should throw instead if `path` is not in the expected format instead of return `bool`.